### DISCUSSION TITLE
OWASP Recommenation for Docker is not to share /var/run/docker.sock with containers

### DISCUSSION THREAD

**Discussion by oussjarrousse** [nginx-proxy/nginx-proxy#2507] [Q&A]
I am using nginx-proxy and the acme-companion, and I really like the control and automation this solution provides.
I even wrote a [blog post about it](https://blog.jarrousse.org/2022/11/01/a-docker-first-solution-to-running-nginx-reverse-proxy-with-automatically-updated-lets-encrypt-ssl-certificates-using-nginx-proxy-and-acme-companion/).

It was brought to my attention that sharing the /var/run/docker.sock goes against the OWASP rules for Docker security:
https://cheatsheetseries.owasp.org/cheatsheets/Docker_Security_Cheat_Sheet.html#rule-1-do-not-expose-the-docker-daemon-socket-even-to-the-containers

What do you recommend here? Is there any way around it?

Maybe it makes sense to add a discussion about the associated risks to the documentation.

Thank you.

**Comment by buchdag** [comment]
Hi @oussjarrousse 

This is a very interesting question, thanks for asking it. I'm no Docker security expert but I'll try to answer as best as I can.

### Short answer

> What do you recommend here ?

- running nginx-proxy in [the separate containers setup](https://github.com/nginx-proxy/nginx-proxy/tree/main/docs#separate-containers), where the Docker daemon socket is mounted in a separate container that do not expose anything to the outside world.
- additionally, [exposing the Docker daemon socket to the docker-gen container on an TLS protected endpoint](https://docs.docker.com/engine/security/protect-access/#use-tls-https-to-protect-the-docker-daemon-socket) rather than using the daemon socket. [docker-gen](https://github.com/nginx-proxy/docker-gen) supports it (see the `DOCKER_HOST` / `DOCKER_CERT_PATH` environment variables). Unfortunately, [acme-companion](https://github.com/nginx-proxy/acme-companion/pull/673) does not (yet, but I'll look into this soon).

> Maybe it makes sense to add a discussion about the associated risks to the documentation.

I completely agree. traefik is in the exact same situation and I quite like [their documentation about this](https://doc.traefik.io/traefik/providers/docker/#docker-api-access). And, like traefik, nginx-proxy can't work without access to the Docker daemon API in one form or in another.

### More detailed answer

Why does nginx-proxy (more specifically docker-gen) need access to the Docker daemon ?
- to get [the system infos](https://docs.docker.com/reference/api/engine/version/v1.47/#tag/System/operation/SystemInfo) from the Docker server.
- to listen (and react) to [the system events](https://docs.docker.com/reference/api/engine/version/v1.47/#tag/System/operation/SystemEvents) happening on the Docker server.
- to get [the list of containers](https://docs.docker.com/reference/api/engine/version/v1.47/#tag/Container/operation/ContainerList) on the Docker server, [the list of networks](https://docs.docker.com/reference/api/engine/version/v1.47/#tag/Network/operation/NetworkList), and [the details for each container](https://docs.docker.com/reference/api/engine/version/v1.47/#tag/Container/operation/ContainerInspect), from which it will render our nginx reverse proxying configuration.
- when nginx and docker-gen run in separate containers, to [send a POSIX SIGHUP signal to the nginx container](https://docs.docker.com/reference/api/engine/version/v1.47/#tag/Container/operation/ContainerKill) when we need the nginx process to reload its config.

As you can see all of those save one are just reading stuff, so what if we could just simply restrict permissions on the Docker daemon socket ?

[Well, no we can't, not simply anyway:](https://docs.docker.com/engine/extend/plugins_authorization/)

> Docker's out-of-the-box authorization model is **all or nothing**. Any user with permission to access the Docker daemon can run any Docker client command. The same is true for callers using Docker's Engine API to contact the daemon. If you require greater access control, you can create authorization plugins and add them to your Docker daemon configuration.

Creating or using Docker authorization plugins is a topic way, way outside the scope of this repository.

So what could we do without taking Docker authorization plugins route ?

1.  ensure that the process that have access to the Docker daemon socket is isolated in its own container that does not expose anything to the outside world. [That's the separate container setup.](https://github.com/nginx-proxy/nginx-proxy/tree/main/docs#separate-containers)
2.  secure the access to the Docker daemon socket. There is two way to do this, [SSH](https://docs.docker.com/engine/security/protect-access/#use-ssh-to-protect-the-docker-daemon-socket) and [TLS](https://docs.docker.com/engine/security/protect-access/#use-tls-https-to-protect-the-docker-daemon-socket). Unfortunately for SSH, docker-gen was created before Docker had an official Go SDK and so had to rely on another project available at that time, [fsouza/go-dockerclient](https://github.com/fsouza/go-dockerclient). TLS support is available with it, but SSH isn't and understandably won't. Re-writing docker-gen to use the official Docker Go SDK isn't out of the table, but that's a big project.
3.  use [tecnativa/docker-socket-proxy](https://github.com/Tecnativa/docker-socket-proxy), the only way that I'm currently aware of to easily control what your service can and cannot do with the Docker API.

Regarding the second point (securing the Docker daemon socket with SSH or TLS) that OWASP insist on : it's not clear to me what difference that would make if the container that have access to the secured endpoint is compromised. AFAIK, if you gain access to this container you also gain access to the SHH or TLS credentials and then to the endpoint, exactly as if the Docker socket was directly mounted to the container. 🤔

I think that **1)** and **3)** are the most immediately important, followed by **2)**, but again I'm no Docker security expert.

**Reply by oussjarrousse** [reply]
Thank you for your comment. Actually, I have already figured out and implemented the solution mentioned in point 1:
https://blog.jarrousse.org/2024/09/01/a-slightly-more-secure-docker-first-solution/
I also wrote a little bit about security issues related to exposing docker sockets to running container facing the public internet. Nothing fancy, just the obvious issues that came to my mind, but maybe it has some added value to the conversation.

**Comment by zeronumbers** [comment]
If you need nginx and cert renewal you can use https://github.com/JonasAlfredsson/docker-nginx-certbot, it doesn't need docker.sock exposed.

**Comment by poehlert** [comment]
I would argue the threat model doesn't change significantly changing from using _local Unix socket_ to TCP based _TLS/SSH locally._ Given the additional complexity and not default behavior, there's not a security win there. If you were already accessing your docker API remotely, then using an encrypted transport is critical but that's not possible with UNIX sockets so changing a local API call to a different local API call buys you almost no benefit and adds a lot of complexity and opportunity to make things worse. The other notes I think are all solid: use separate containers so internet facing ones don't have docker API access and add an AuthZ plugin if you really think you need it, though is pretty advanced. 

For hardening some other good ones that get you pretty far are running nginx as non-root and providing read-only file systems to limit living off the land if someone does break in. Active monitoring and hardening your origin servers can also be done if you need to secure your deployment.
